java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateMethodCheck.java

```diff
@@ -82,7 +82,11 @@ private static Stream<MethodTree> findUnusedPrivateMethods(ClassTree tree) {
     methodNames.removeAll(collector.unresolvedMethodNames);
     var filter = new MethodsUsedInAnnotationsFilter(methodNames);
     tree.accept(filter);
-    return unusedPrivateMethods.stream().filter(it -> filter.filteredNames.contains(it.simpleName().name()));
+    var methodTreeStream = unusedPrivateMethods.stream().filter(it -> filter.filteredNames.contains(it.simpleName().name()));
+    var annotatedMethodCollector = new AnnotatedMethodCollector();
+    tree.members().forEach(it -> it.accept(annotatedMethodCollector));
+    var methodSourceAnnotatedMethods = annotatedMethodCollector.methodSourceAnnotatedMethods;
+    return methodTreeStream.filter(it -> !methodSourceAnnotatedMethods.contains(it.simpleName().name()));
```

Comment:
I think we might incur in new false negatives with this logic, if it's the unused private method itself to have the `@MethodSource` annotation (which you might argue does not make sense in most cases):
```
  @MethodSource
  private void unusedMethodWithWrongAnnotation(){} // FN
```

You could encounter unused private methods which mistakenly (or not) have that annotation, and they would not raise an issue

Reply:
Hmm. On line 138 there is a check:
`if (isUnusedPrivate(symbol) && hasNoAnnotation(methodTree) `
So, we should be safe.

Reply:
I think that helps prevent false positives, but with the filtering we do before, around lines ~80, methods annotated with `@MethodSource` will never reach the reporting part.
So if you try adding the example I pasted above, you should notice that no issue is raised for that method, even if is in fact private and unused.